### PR TITLE
NN-3363: Improvements to new CSRA ratings endpoint

### PR DESCRIPTION
**Pull request by sp-watson**
Batch DB calls for CSRA ratings for multiple offenders. Also only return data for offenders that have assessments

### CODE REVIEW

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAssessmentService.java [hunk 1]
```diff
@@ -120,20 +124,39 @@ private AssessmentQuestion getAssessmentQuestionAndAnswers(final AssessmentEntry
             answers.stream().skip(1).collect(toList()));
     }
 
-    public List<AssessmentClassification> getOffendersAssessmentRatings(List<String> offenderList) {
-        return offenderList.stream().map(this::getOffenderCurrentAssessmentRating).collect(Collectors.toList());
+    @Transactional(readOnly = true)
+    public List<AssessmentClassification> getOffendersAssessmentRatings(final List<String> offenderNos) {
+        final var classifications = new ArrayList<AssessmentClassification>();
+        final var batch = Lists.partition(new ArrayList<>(offenderNos), maxBatchSize);
+        batch.forEach(offenderNosBatch -> classifications.addAll(getOffendersCurrentAssessmentRating(offenderNosBatch)));
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAssessmentService.java:131]: Meh, could be a flatMap to keep up with consistency

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAssessmentService.java [hunk 2]
```diff
@@ -120,20 +124,39 @@ private AssessmentQuestion getAssessmentQuestionAndAnswers(final AssessmentEntry
...
+            .map(e -> getOffenderCurrentAssessmentRating(e.getKey(), e.getValue()))
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .collect(toList());
     }
 
-    private AssessmentClassification getOffenderCurrentAssessmentRating(String offenderNo) {
-        var currentClassification = getCurrentCsraClassification(offenderNo);
+    private Optional<AssessmentClassification> getOffenderCurrentAssessmentRating(final String offenderNo, final List<OffenderAssessment> assessmentsLatestFirst) {
+        var currentClassification = calculateCurrentCsraClassification(assessmentsLatestFirst);
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAssessmentService.java:146]: Prob needs to be final (sorry :runner: )

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAssessmentService.java [hunk 3]
```diff
@@ -120,20 +124,39 @@ private AssessmentQuestion getAssessmentQuestionAndAnswers(final AssessmentEntry
...
-        return AssessmentClassification.builder()
+        return Optional.of(AssessmentClassification.builder()
             .offenderNo(offenderNo)
             .classificationCode(currentClassification.classificationCode)
             .classificationDate(currentClassification.classificationDate)
-            .build();
+            .build());
+    }
+
+    private CurrentCsraAssessment calculateCurrentCsraClassification(final List<OffenderAssessment> offenderAssessmentsLatestFirst) {
```
Comment (JamieDraperUK) [inline src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAssessmentService.java:157]: Would be be better to return a Optional here?

**Review by JamieDraperUK — DISMISSED** [review]
LGTM

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAssessmentService.java (hunk 1 above)
Comment (sp-watson) [reply src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAssessmentService.java:131]: Yes, that was the comment you had before and i fixed it... I'll do it again

File: src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAssessmentService.java (hunk 3 above)
Comment (sp-watson) [reply src/main/java/uk/gov/justice/hmpps/prison/service/OffenderAssessmentService.java:157]: I spent about 5 minutes thinking about this and decided to leave as is. The other method that does use Optional was done so we could better chain calls in a stream.

**Review by JamieDraperUK — APPROVED** [review]
LGTM

**Pull request merged**